File: solutions/ruby/pig-latin/1/pig_latin.rb
module PigLatin
  def self.translate(phrase)
    phrase.split.map { |word| translate_word(word) }.join(' ')
  end

  def self.translate_word(word)
    if word =~ /\A([aeiou]|xr|yt)/
      word + 'ay'
    elsif word =~ /\A([^aeiou]*qu)(.*)\z/
      $2 + $1 + 'ay'
    elsif word =~ /\A([^aeiou]+)(y.*)\z/
      $2 + $1 + 'ay'
    elsif word =~ /\A([^aeiou]+)(.*)\z/
      $2 + $1 + 'ay'
    else
      word + 'ay'
    end
  end
end

File: solutions/javascript/pig-latin/1/pig-latin.js
const VOWELS = 'aeiou';

function translateWord(word) {
  if (/^[aeiou]|^xr|^yt/.test(word)) {
    return word + 'ay';
  }

  let cluster = '';
  let i = 0;

  if (word[0] === 'y') {
    cluster += 'y';
    i = 1;
  }

  while (i < word.length) {
    if (word.slice(i, i + 2) === 'qu') {
      cluster += 'qu';
      i += 2;
    } else if (word[i] === 'y') {
      break;
    } else if (!VOWELS.includes(word[i])) {
      cluster += word[i];
      i++;
    } else {
      break;
    }
  }

  return word.slice(i) + cluster + 'ay';
}

export const translate = (phrase) => {
  return phrase.split(' ').map(translateWord).join(' ');
};
